bazel/cas/service.go

```diff
@@ -66,27 +98,173 @@ func (s *casServer) GetTree(
 
 // Serves content in the bundlestore to a client via grpc streaming.
 // Implements googleapis bytestream Read
-func (s *casServer) Read(req *googlebytestream.ReadRequest, ser googlebytestream.ByteStream_ReadServer) error {
+func (s *casServer) Read(req *bytestream.ReadRequest, ser bytestream.ByteStream_ReadServer) error {
 	log.Infof("Received CAS Read request: %s", req)
-	// Real implementation: fetch resource from backend and call Send(Data []byte) until finished
+
+	if !s.IsInitialized() {
+		return status.Error(codes.Internal, "Server not initialized")
+	}
+
+	// Parse resource name per Bazel API specification
+	resource, err := ParseReadResource(req.GetResourceName())
+	if err != nil {
+		return status.Error(codes.InvalidArgument, fmt.Sprintf("%v", err))
+	}
+
+	// Input validation per API spec
+	if req.GetReadOffset() < 0 || req.GetReadOffset() >= resource.Digest.GetSizeBytes() {
```

Comment:
In line with my other comment about size, can you drop the GetReadOffset() >= resource.Digest.GetSizeBytes() check (and make the check be against the actual underlying blob's real size)?